README.md

```diff
@@ -7,185 +7,181 @@
 
-### Example
+## Code Generation
+
+The code generator generates parsers based on message specifications. The generated parser allows to validate and dissect messages and thereby respects all specified restrictions between message fields and related messages. Adding the generation of messages is in planning. By using [SPARK](https://www.adacore.com/about-spark) we are able to prove the absence of runtime errors and prevent the incorrect usage of the generated code (e.g., enforce that a field of a message is validated before accessed).
+
+The code generator creates a number of packages for a specification. All basic types like integers, enumerations and arrays are collectively declared in one package. For each message a child package is generated which contains validation and access functions for every field of the message.
+
+A user of the generated code has to validate a message field or the whole message before accessing the data of a particular message field. The SPARK verifcation tools in combination with the generated contracts make it possible to ensure this property, and so prevent incorrect usage.
+
+## Usage
 
-The following sample specification describes a protocol `Foo` with one protocol message type `Bar`. Three fields are specified for `Bar`:
+The `rflx` tool is used to verify a specification and generate code based on it. Therefore it offers the two sub-commands `check` and `generate`.
 
-- A `Tag` field with 2 bit,
-- a `Value_Length` field with 14 bit, and
-- a `Value` field, whose length is specified by the value in `Value_Length`.
+## Example
 
-The `Tag` can have two valid values: `1` (`Msg_Data`) and `3` (`Msg_Error`). In the case `Tag` has a value of `1` the fields `Value_Length` and `Value` follow. `Bar` contains only one field, if the value of `Tag` is `3`. All other values of `Tag` lead to an invalid message.
+In the following, the complete process of specifying a message, generating code, and using the generated code is demonstrated using a small example.
+
+### Specification
+
+The following sample specification describes a protocol `TLV` with one message type `Message` consisting of three fields:
+
+- A field `Tag` of 2 bit length,
+- a field `Value_Length` of 14 bit length, and
+- a field `Value`, whose length is specified by the value in `Value_Length`.
+
+The `Tag` can have two valid values: `1` (`Msg_Data`) and `3` (`Msg_Error`). In case `Tag` has a value of `1` the fields `Value_Length` and `Value` follow. `Message` contains only one field, if the value of `Tag` is `3`. All other values of `Tag` lead to an invalid message.
 
 ```
-package Foo is
+package TLV is
 
    type Tag_Type is (Msg_Data => 1, Msg_Error => 3) with Size => 2;
    type Length_Type is mod 2**14;
 
-   type Bar is
+   type Message is
       message
-         Tag          : Tag_Type
-            then Value_Length
+         Tag    : Tag_Type
+            then Length
```

Comment:
We should at least mention the graph structure of the spec and what `then` is good for in the intro above. Otherwise, I think it is hard to grasp what the example specification is intended to express.